Utility mixin additions

mdo opened the pull request:
Builds on #34102 to also fix #33861. Key changes from original PR:

- Rearrange some source files
- Document the new `utl()` mixin, the motivation, and examples

Todo:

- [x] Remove test changes from `bootstrap.scss`
- [x] Iron out details around media queries and document it
- [x] Rename from `utl()` to `util()`?
- [x] Add docs for extending utilities via API and then using those new utils via mixin

/cc @donquixote

donquixote committed:
Issue #33861: New utl() mixin

Co-Authored-By: Andreas Hennings <andreas@dqxtech.net>

ffoodd reviewed (commented):
A few sentences feel a bit weird to me, might need somme clarification. I'll try to read the docs again later to make some suggestions if needed.

The overall PR looks good, and I like the feature. Regarding the mixin name, I'd be in favor of `util()` and even `utility()`.

We could also consider using `apply()` as it's pretty close to the named Tailwind feature.

donquixote commented on scss/functions/_utilities-map.scss:
```diff
@@ -1,17 +1,19 @@
   }
 
-  $utilities-map: ();
+  $utilities-map: () !default;
```
Is adding `!default` also part of the "linter" operation?
I usually try to have an isolated commit where I _only_ do style and whitespace changes, that do not change functionality.
This change does change functionality - in a good way, I would think. So could be in separate commit.

donquixote commented:
@mdo For my taste you could rearrange the commits a bit, e.g.:
- squash the last fix (removing leftover code) into the commit where this was introduced
- split the `!default` out of the "Linter" commit, so that only non-functional changes remain.
- Dedicated commit for introducing the docs?
- Only change your own commits, don't touch the first one. This way we still see who did what.

But I don't know if this is something you commonly do. For some people, rebase + force-push is blasphemy, and ofc it can mess with others who try to work on the same PR.

donquixote commented on site/content/docs/5.1/utilities/mixin.md:
```diff
@@ -0,0 +1,83 @@
+
+There are two major motivations for creating the utility mixin:
+
+1. **Bootstrap should be as approachable and useful to as many people as possible.** This is why we provide compiled ready-to-go distribution files alongside source Sass and JavaScript. Similarly, not everyone likes pre-built components, and not everyone likes completely utility-driven development.
```
Just saying, the choice is not always based on personal preference.
Often you have constraints in the project or ecosystem you work with. E.g. your web framework or CMS might make it difficult to change some of the html, or the html is coming from a 3rd party. So you need to do everything in CSS instead.

donquixote commented:
Something I don't remember: Is it possible to avoid generating the utility classes, but still use the mixins?
E.g. if you don't control any of your html, you probably want to completely skip utilities and only use the mixins. Removing utility classes then makes the CSS smaller.
If this is something that can be done, it should be mentioned in the docs (perhaps it is already explained elsewhere, then we just need a link).

mdo commented:
> Something I don't remember: Is it possible to avoid generating the utility classes, but still use the mixins?

In my brief testing, yes, you can comment out `@include generate-utility()` instances in `scss/utilities/_api.scss` and regenerate the CSS. It won't include any utility classes, but will allow you to consume the placeholders. Needs more testing to be sure, but seemed to work. Might be something we add with a Sass-variable feature flag like `$enable-utility-classes`.

mdo committed:
Update utility mixin and wrap utility classes in `$enable-` variable

- Rearrange Sass files to simplify things
- Rename `utl()` to `util()`
- Add new `$enable-utility-classes` variable for disabling the default generation of our utilities (useful if you want to only use utilities via mixin)

mdo committed:
Document the `util()` mixin

WinterSilence commented:
why not use native `@extend .class-name`? 
why multiple class definition not supported? `[@mixin util($classes...) {}`](https://sass-lang.com/documentation/at-rules/mixin#taking-arbitrary-arguments)
this mixin no supports [`null` key  in `value` map](https://github.com/twbs/bootstrap/blob/28f7c94475b9456897ef9a4b60bc0f45d4db3388/scss/_utilities.scss#L613) or [`class: null`](https://github.com/twbs/bootstrap/blob/main/scss/_utilities.scss#L617)

WinterSilence commented on scss/functions/_utilities-map.scss:
```diff
@@ -0,0 +1,89 @@
+  $utilities-map: () !default;
+
+  @each $key, $utility in $utilities {
+    @if type-of($utility) == "map" {
```
[type-of($value) //=> unquoted string](https://sass-lang.com/documentation/modules/meta#type-of)

mdo replied:
We're inconsistent in this across our source code, so going to leave as-is for now.

WinterSilence commented on scss/functions/_utilities-map.scss:
```diff
@@ -0,0 +1,89 @@
+      $breakpoints-map: map-merge(
+        $breakpoints-map,
+        (
+          "-" + $name: $name,
```
SASS developers recommend use  interpolation `-#{$name}` instead concatenation because it's work more stable

mdo replied:
Done!

mdo committed:
Fix up docs

mdo committed:
Add ability to use multiple classes in a single mixin call

mdo commented:
> why not use native `@extend .class-name`?

In general, we shy away from `@extend` as it leads to an excessive number of selectors.

> why multiple class definition not supported? [`@mixin util($classes...) {}`](https://sass-lang.com/documentation/at-rules/mixin#taking-arbitrary-arguments)

Good call! I just pushed a fix to make that happen.

> this mixin no supports [`null` key in `value` map](https://github.com/twbs/bootstrap/blob/28f7c94475b9456897ef9a4b60bc0f45d4db3388/scss/_utilities.scss#L613) and [`class: null`](https://github.com/twbs/bootstrap/blob/main/scss/_utilities.scss#L617)

This mixin is intended to give you an option to use the classes that are generated, not to generate new classes for you. That'll still happen in the API and that giant Sass map we have.

mdo committed:
Document how to disable generated classes and still use the util() mixin

mdo committed:
interpolate instead

mdo commented:
This is ready to go @twbs/css-review. I'm down to get it into v5.3.0 stable, or make this part of a v5.4.0 release (which I would theorize as being the final v5 minor release before v6.

julien-deramond committed:
Merge branch 'main' into pr/34102

julien-deramond commented:
I haven't reviewed the entire PR but just played with it.
I might have found an issue.

```css
.badge {
  &.test {
    @include util(p-3);
    @include util(mb-3);
    @include util(bg-primary);
    @include util(text-primary);
    @include util(text-break);
  }
}
```

First, IDK if we will ever have to use this `util()` ourselves in Bootstrap, but if so, `@import "utilities/api";` is included in `bootstrap.scss` file in the end, so if we want to use it to define badges for instance, it won't compile.

The other thing I've spotted is when you use `@include util(text-primary)` or `@include util(bg-primary)`.
The generated CSS will be

```css
.badge.test {
  padding: 1rem;
  margin-bottom: 1rem;
  background-color: rgba(var(--bs-primary-rgb), var(--bs-bg-opacity));
  color: rgba(var(--bs-primary-rgb), var(--bs-text-opacity));
  word-wrap: break-word;
  word-break: break-word;
}
```

And as you can see, for `bg-primary` and `text-primary`, `--bs-bg-opacity: 1;` and `--bs-text-opacity: 1;` are missing so it doesn't work.

As a reminder

```css
.text-primary {
  --bs-text-opacity: 1;
  color: rgba(var(--bs-primary-rgb), var(--bs-text-opacity)) !important;
}

.bg-primary {
  --bs-bg-opacity: 1;
  background-color: rgba(var(--bs-primary-rgb), var(--bs-bg-opacity)) !important;
}
```

louismaximepiton reviewed (commented):
Should we handle the stateful classes and utilities with only CSS var?

For instance, 
- `util(focus-ring)` raises the following: `scss\utilities\_api.scss:73 Debug: Unknown utility class focus-ring`.
- `util(focus-ring-primary)` raises the following:
```sh
CssSyntaxError: C:\Applications\bootstrap\scss\utilities\_api.scss:68:10: Unknown word

  66 |       @else {
  67 |         @each $property in map-get($definition, properties) {
> 68 |           #{$property}: map-get($definition, value);
     |          ^
  69 |         }
  70 |       }
```

mdo commented:
Closing for #42630.
